### PR TITLE
682 data types

### PR DESCRIPTION
**Pull request by alvagante**
Checked syntax. Not spec tests.
Some choices are worth discussion.

### CODE REVIEW

File: spec/classes/sensu_init_spec.rb [hunk 1]
```diff
@@ -402,7 +402,7 @@
           context "when #{var_name} (#{type}) is set to invalid #{invalid} (as #{invalid.class})" do
             let(:params) { [mandatory_params, var[:params], { :"#{var_name}" => invalid, }].reduce(:merge) }
             it 'should fail' do
-              expect { should contain_class(subject) }.to raise_error(Puppet::Error, /#{var[:message]}/)
+              expect { should contain_class(subject) }.to raise_error(Puppet::PreformattedError)
```
Comment (Phil-Friderici) [inline spec/classes/sensu_init_spec.rb]: This would remove the test for the specific error messages provided in the hash as 'message'.
Please don't change.

**Comment by Phil-Friderici** [comment]
Good to see this :)

File: manifests/check.pp [hunk 2]
```diff
@@ -129,89 +129,39 @@
...
-    fail("sensu::check{${name}}: subscribers must be an array or a string (got: ${subscribers})")
-  }
-  if $subdue {
-    if is_hash($subdue) {
-      if !( has_key($subdue, 'days') and is_hash($subdue['days']) ){
-        fail("sensu::check{${name}}: subdue hash should have a proper format. (got: ${subdue}) See https://sensuapp.org/docs/latest/reference/checks.html#subdue-attributes")
-      }
-    } elsif !($subdue == 'absent') {
-      fail("sensu::check{${name}}: subdue must be a hash or 'absent' (got: ${subdue})")
+  if $subdue =~ Hash {
```
Comment (ghoneycutt) [inline manifests/check.pp]: do we need a struct for this?

File: manifests/init.pp [hunk 3]
```diff
@@ -445,153 +445,130 @@
 #   Default: 'Sensu'
 
 class sensu (
-  $version                        = 'installed',
-  $sensu_plugin_name              = 'sensu-plugin',
-  $sensu_plugin_provider          = $::osfamily ? {
+  Pattern[/^absent$/, /^installed$/, /^latest$/, /^present$/, /^[\d\.\-el]+$/] $version = 'installed',
```
Comment (ghoneycutt) [inline manifests/init.pp]: Not sure we should be so picky here. Perhaps `String` is right here. What do you think?

File: spec/classes/sensu_package_spec.rb [hunk 4]
```diff
@@ -294,7 +294,7 @@
         let(:params) { { :purge => 'a_string' } }
 
         it 'should fail' do
-          expect { should create_class('sensu') }.to raise_error(/not a Hash/)
+          expect { should create_class('sensu') }.to raise_error(Puppet::PreformattedError)
```
Comment (ghoneycutt) [inline spec/classes/sensu_package_spec.rb]: I'm not familiar with `Puppet::PreformattedError` what is this?

File: manifests/check.pp (hunk 2 above)
Comment (alvagante) [inline manifests/check.pp]: Here yes, because value can also be the string 'absent' and this check has to be done only on hashes.

File: manifests/init.pp (hunk 3 above)
Comment (alvagante) [inline manifests/init.pp]: Some tests might fail, and this is probably the same validation which was there before

File: spec/classes/sensu_package_spec.rb (hunk 4 above)
Comment (alvagante) [inline spec/classes/sensu_package_spec.rb]: The error we were having in failing spec tests. Related to type validation

File: manifests/check.pp [hunk 5]
```diff
@@ -129,89 +129,39 @@
...
-  $proxy_requests      = undef,
+  String $command,
+  Enum['present','absent'] $ensure              = 'present',
+  Optional[String] $type                = undef,
+  Variant[Undef,String,Array] $handlers            = undef,
+  Optional[Array] $contacts            = undef,
+  Variant[Boolean,Enum['absent']] $standalone          = true,
+  String $cron                = 'absent',
+  Variant[Integer,Enum['absent']] $interval            = 60,
+  Variant[Undef,Pattern[/^(\d+)$/],Integer,Enum['absent']] $occurrences         = undef,
```
Comment (alvagante) [inline manifests/check.pp]: RFC: In the module in most parts, numbers as accepted both as string and numbers. I've used this approach (Pattern[/^(\d+)$/],Integer ) in these cases, there could be something better maybe.

Comment (ghoneycutt) [inline manifests/check.pp]: Asking for clarification in related ticket #582

File: spec/classes/sensu_init_spec.rb (hunk 1 above)
Comment (alvagante) [inline spec/classes/sensu_init_spec.rb]: Oops, looks my my reply here got lost... 
so yes, I should update the message texts and check on them.
Wonder it it's actually really necessary as they are mostly normal Puppet type validation error messages, so not actually under the module code control.

File: manifests/check.pp (hunk 2 above)
Comment (ghoneycutt) [inline manifests/check.pp]: The module has many type/providers though we are looking to minimize the number through using sensu_sorted_json().

Comment (alvagante) [inline manifests/check.pp]: Sorry, I was meaning data types, to write in the types dir

Comment (ghoneycutt) [inline manifests/check.pp]: If you think that adding module specific data types will simplify things, then let's do it.

File: spec/classes/sensu_init_spec.rb (hunk 1 above)
Comment (alvagante) [inline spec/classes/sensu_init_spec.rb]: @Phil-Friderici I couldn't do any better that this.
To match the actual error messages according to data type would need a refactoring of the test code.... maybe later ;-)

**Comment by alvagante** [comment]
So @ghoneycutt from my pov the PR is basically ready.
Some choices have been done which need review:
- How to manage numbers.
Current approach accepts digits both passed as strings (but still digits) or Numeric.
This is needed to pass some of the existing tests, which would fail otherwise.
Current behaviour is somehow more permissive and this is better, IMHO, at is limits the risk to avoid breaking existing setups.
- On the same spirit, In various cases the exact accepted type of parameters was not clean (to me), generally I've tried to follow the less restrictive approach. (Any)
- Verification of the error message on some input validation checks have been simplified/generalized to avoid the need to refactor whole spec classes.  

Needless to say that whatever the tests may cover, with such an invasive commit we may break existing setups using some edge case not tested.

File: spec/classes/sensu_init_spec.rb (hunk 1 above)
Comment (Phil-Friderici) [inline spec/classes/sensu_init_spec.rb]: @alvagante I still like to test that input parameter validation is as expected. Think this makes even sense with Puppet v4 and the new data type validations. It is not mandatory to have validation in Puppet v4 so it's still a module feature to take care for.

Testing for the actual error messages does proof that there is no other error popping up. With the general ```PreformattedError``` it would accept any error message regardless of the real reason.

File: spec/classes/sensu_package_spec.rb (hunk 4 above)
Comment (Phil-Friderici) [inline spec/classes/sensu_package_spec.rb]: ```Puppet::PreformattedError``` will accept any error thrown, independently from the reason for the fail. The intention here is to ensure that the module really fails on the validation errors.

**Comment by alvagante** [comment]
@ghoneycutt  I squashed this one in a single commit.
I'll leave the Sensitive part the the relevant PR, to merge once this one is merged.

Can eventually fix later some details like the ones based on pending decisions, or the ones pointed out by @Phil-Friderici  and @jeffmccune 

I'd prefer to merge and consolidate asap a common code base based on puppet 4 types.

**Comment by ghoneycutt** [comment]
Released in v2.26.0
